components/datetime/src/raw/datetime.rs

```diff
@@ -58,11 +61,29 @@ impl DateTimeFormat {
             calendar,
         )?;
 
-        let requires_data = datetime::analyze_patterns(&patterns.get().0, false)
+        let required = datetime::analyze_patterns(&patterns.get().0, false)
             .map_err(|field| DateTimeFormatError::UnsupportedField(field.symbol))?;
 
         let langid: icu_locid::LanguageIdentifier = locale.clone().into();
 
+        let week_data = if required.week_data {
+            Some(
+                data_provider
+                    .load_payload(&DataRequest {
+                        resource_path: ResourcePath {
+                            key: provider::key::WEEK_DATA_V1,
+                            options: ResourceOptions {
+                                variant: langid.region.map(|r| r.as_str().to_string().into()),
+                                langid: None,
+                            },
```

Comment:
This is the first time in ICU4X where we are adding *regional* data.  How do we best express that in the resource path?  This is the main architectural question of this PR.

Overall, I do like smaller data keys and hoisting as much of the data as possible into a layer visible to the data provider. This helps with data size and data slicing.

However, I'm unsure about using the `variant` field rather than the `langid` field (with `und`, like `und-MO`).  If we used `langid`, it would be easier to programatically perform data slicing of certain regions.  A likely downside is that `und-MO` is more bytes than just `MO`.

Interested to hear @zbraniecki's thoughts on how to represent regional data in the provider.

Reply:
I did briefly try using und-${REGION}. IIRC my main concern was that it made datagen config a bit akward:
1. You'd have to request both ${LANG}-${REGION} and und-${REGION} (at least until #834 IIUC).
2. There are valid lang-region pairs that are not defined in CLDR but that are valid for weekdata (e.g. fr & FR both exist but not fr-FR). In unit tests at least 'fr-FR' does not fallback to 'fr' for non weekdata leading to unexpected failures (by adding 'und-FR' to datagen one would expect *-FR to work). I haven't checked if this is test specific (seen in fixture tests using icu_testdata::get_provider() rather than the unit tests using MultiKeyStructProvider).